lib/index.js

```diff
@@ -2,115 +2,113 @@
-    pkg: require('../package.json')
 };
 
 function isFunction(functionToCheck) {
-  return Object.prototype.toString.call(functionToCheck) === '[object Function]';
+  const objectProto = Object.prototype.toString.call(functionToCheck);
+  return objectProto === '[object Function]' || objectProto === '[object AsyncFunction]';
 }
 
 internals.implementation = function (server, options) {
 
   var validationResult = Joi.validate(options, optionsSchema);
-
-  if (validationResult.error){
+  if (validationResult.error) {
     throw new Error(validationResult.error.message);
   }
 
   var settings = Hoek.clone(options);
 
   var scheme = {
-    authenticate: function (request, reply) {
+    authenticate: async function (request, h) {
 
       var req = request.raw.req;
       var authorization = req.headers.authorization;
       if (!authorization) {
-        return reply(Boom.unauthorized(null, 'Bearer'));
+        throw Boom.unauthorized(null, 'Bearer');
       }
 
       var parts = authorization.split(/\s+/);
 
       if (parts.length !== 2) {
-        return reply(Boom.badRequest('Bad HTTP authentication header format', 'Bearer'));
+        throw Boom.badRequest('Bad HTTP authentication header format', 'Bearer');
       }
 
       if (parts[0].toLowerCase() !== 'bearer') {
-        return reply(Boom.unauthorized(null, 'Bearer'));
+        throw Boom.unauthorized(null, 'Bearer');
       }
 
-      if(parts[1].split('.').length !== 3) {
-        return reply(Boom.badRequest('Bad HTTP authentication header format', 'Bearer'));
+      if (parts[1].split('.').length !== 3) {
+        throw Boom.badRequest('Bad HTTP authentication header format', 'Bearer');
       }
 
       var token = parts[1];
 
       var getKey = isFunction(settings.key) ?
         settings.key :
-        function(req, token, callback) { callback(null, settings.key); };
+        function (req, token) { return { key: settings.key }; };
 
-      getKey(request, token, function(err, key, extraInfo){
-        if (err) { return reply(Boom.wrap(err)); }
-        // handle err
-        jwt.verify(token, key, settings, function(err, decoded) {
-          
-          if(err) {
-            return reply(Boom.unauthorized( 'JSON Web Token validation failed: ' + err.message, 'Bearer'));
-          }
-
-          if (!settings.validateFunc) {
-            return reply.continue({ credentials: decoded });
-          }
-
-          settings.validateFunc(decoded, extraInfo, function (err, isValid, credentials) {
-
-            credentials = credentials || null;
+      let keyResult;
+      try {
+        keyResult = await getKey(request, token);
+      } catch (err) {
+        throw err;
+      }
+      const { key, extraInfo } = keyResult;
 
-            if (err) {
-              return reply(err, null, { credentials: credentials, log: { tags: ['auth', 'jwt'], data: err } });
-            }
+      let decoded;
+      try {
+        decoded = jwt.verify(token, key, settings);
```

Comment:
I guess you have consider this https://github.com/auth0/node-jsonwebtoken/pull/285 when doing this change, if this is true for this version of jwt lib then 👍 otherwise if we are making sync something that is actually async it might generate a big performance degradation and auth tends to be in all critical paths so if it is not already an sync operation hidden by the callback I would reconsider this change. (Might wort documenting the reasons for the decision too, as a comment or so)

Reply:
For what I've seen, `verify` is really sync, so this seems like the best approach.

Reply:
agree